NeoML/src/TraditionalML/Svm.cpp

```diff
@@ -22,15 +22,144 @@ limitations under the License.
 #include <NeoML/TraditionalML/PlattScalling.h>
 #include <SvmBinaryModel.h>
 #include <LinearBinaryModel.h>
-#include <NeoMathEngine/OpenMP.h>
+#include <NeoMathEngine/ThreadPool.h>
 #include <SMOptimizer.h>
 
 namespace NeoML {
 
+namespace {
+
+// Function to decide run the task in a single thread or in parallel
+static inline bool isThreadTaskRelevant( int64_t operationCount )
+{
+	constexpr int64_t MinOmpOperationCount = 32768;
+	return operationCount >= MinOmpOperationCount;
+}
+
+// Task which processes in multiple threads
+struct IThreadTask {
+	virtual ~IThreadTask() {}
+
+	void ParallelRun();
+protected:
+	IThreadTask( IThreadPool& threadPool, const IProblem& );
+
+	int ParallelizeSize() const { return Problem.GetVectorCount(); }
+	int ThreadCount() const { return ThreadPool.Size(); }
+	void RunSplittedByThreads( int threadIndex );
+	void Run( int threadIndex, int index, int count );
+
+	virtual void RunOnElement( int threadIndex, int index, const CFloatVectorDesc& ) = 0;
+
+	IThreadPool& ThreadPool;
+	const IProblem& Problem;
+	const CFloatMatrixDesc Matrix;
+};
+
+IThreadTask::IThreadTask( IThreadPool& threadPool, const IProblem& problem ) :
+	ThreadPool( threadPool ),
+	Problem( problem ),
+	Matrix( Problem.GetMatrix() )
+{
+	NeoAssert( Matrix.Height == Problem.GetVectorCount() );
+	NeoAssert( Matrix.Width == Problem.GetFeatureCount() );
+}
+
+void IThreadTask::ParallelRun()
+{
+	if( !isThreadTaskRelevant( Problem.GetVectorCount() ) ) {
```

Comment:
See [this](https://github.com/neoml-lib/neoml/pull/877#discussion_r1244001799)

Reply:
fixed